// @databricks/lakebase-eaa4c28d5584f767/dist/config.js
import { ConfigurationError, ValidationError } from "./errors.js";
import { WorkspaceClient } from "@databricks/sdk-experimental";

//#region src/config.ts
/** Default configuration values for the Lakebase connector */
const defaults = {
	port: 5432,
	sslMode: "require",
	max: 10,
	idleTimeoutMillis: 3e4,
	connectionTimeoutMillis: 1e4
};
const VALID_SSL_MODES = [
	"require",
	"disable",
	"prefer"
];
/** Parse pool configuration from provided config and environment variables */
function parsePoolConfig(userConfig) {
	const endpoint = userConfig?.endpoint ?? process.env.LAKEBASE_ENDPOINT;
	if (!endpoint && !userConfig?.password) throw ConfigurationError.missingEnvVar("LAKEBASE_ENDPOINT or config.endpoint (or provide config.password for native auth)");
	const host = userConfig?.host ?? process.env.PGHOST;
	if (!host) throw ConfigurationError.missingEnvVar("PGHOST or config.host");
	const database = userConfig?.database ?? process.env.PGDATABASE;
	if (!database) throw ConfigurationError.missingEnvVar("PGDATABASE or config.database");
	const portStr = process.env.PGPORT;
	const port = userConfig?.port ?? (portStr ? Number.parseInt(portStr, 10) : defaults.port);
	if (Number.isNaN(port)) throw ValidationError.invalidValue("port", portStr, "a number");
	const sslMode = validateSslMode(userConfig?.sslMode ?? process.env.PGSSLMODE ?? void 0) ?? defaults.sslMode;
	const max = userConfig?.max ?? defaults.max;
	const idleTimeoutMillis = userConfig?.idleTimeoutMillis ?? defaults.idleTimeoutMillis;
	const connectionTimeoutMillis = userConfig?.connectionTimeoutMillis ?? defaults.connectionTimeoutMillis;
	return {
		endpoint,
		host,
		database,
		port,
		sslMode,
		ssl: userConfig?.ssl,
		max,
		idleTimeoutMillis,
		connectionTimeoutMillis
	};
}
/** Validate and return the SSL mode, or undefined when not set */
function validateSslMode(value) {
	if (value === void 0) return;
	if (!VALID_SSL_MODES.includes(value)) throw ValidationError.invalidValue("sslMode (PGSSLMODE)", value, `one of: ${VALID_SSL_MODES.join(", ")}`);
	return value;
}
/** Get workspace client from config or SDK default auth chain */
function getWorkspaceClient(config) {
	if (config.workspaceClient) return config.workspaceClient;
	return new WorkspaceClient({});
}
/** Get username synchronously from config or environment */
function getUsernameSync(config) {
	if (config.user) return config.user;
	const pgUser = process.env.PGUSER;
	if (pgUser) return pgUser;
	const clientId = process.env.DATABRICKS_CLIENT_ID;
	if (clientId) return clientId;
	throw ConfigurationError.missingEnvVar("config.user, PGUSER or DATABRICKS_CLIENT_ID");
}
/**
* Resolves the PostgreSQL username for a Lakebase connection.
*
* Extends {@link getUsernameSync} with an async fallback that fetches the
* current user's identity from the Databricks workspace API. Use this when
* you don't have an explicit username configured and want automatic resolution
* (e.g. human users authenticating via PAT or browser OAuth in ~/.databrickscfg).
*
* Resolution priority:
* 1. `config.user` — explicit config value
* 2. `PGUSER` env var
* 3. `DATABRICKS_CLIENT_ID` env var (service principals)
* 4. `currentUser.me()` — fetched from Databricks workspace API
*
* Returns `undefined` if all attempts fail rather than throwing, so the
* caller can decide whether to proceed or surface an error.
*/
async function getUsernameWithApiLookup(config) {
	try {
		return getUsernameSync(config ?? {});
	} catch {}
	try {
		return (await getWorkspaceClient(config ?? {}).currentUser.me()).userName ?? void 0;
	} catch {
		return;
	}
}

//#endregion
export { getUsernameSync, getUsernameWithApiLookup, getWorkspaceClient, parsePoolConfig };
