// @databricks/lakebase-eaa4c28d5584f767/dist/errors.js
//#region src/errors.ts
/**
* Base error class for Lakebase driver errors.
*/
var LakebaseError = class extends Error {
	cause;
	context;
	constructor(message, options) {
		super(message);
		this.name = this.constructor.name;
		this.cause = options?.cause;
		this.context = options?.context;
		if (Error.captureStackTrace) Error.captureStackTrace(this, this.constructor);
	}
};
/**
* Error thrown when configuration is missing or invalid.
*/
var ConfigurationError = class ConfigurationError extends LakebaseError {
	code = "CONFIGURATION_ERROR";
	/**
	* Create a configuration error for missing environment variable
	*/
	static missingEnvVar(varName) {
		return new ConfigurationError(`${varName} environment variable is required`, { context: { envVar: varName } });
	}
};
/**
* Error thrown when input validation fails.
*/
var ValidationError = class ValidationError extends LakebaseError {
	code = "VALIDATION_ERROR";
	/**
	* Create a validation error for an invalid field value
	*/
	static invalidValue(fieldName, value, expected) {
		return new ValidationError(expected ? `Invalid value for ${fieldName}: expected ${expected}` : `Invalid value for ${fieldName}`, { context: {
			field: fieldName,
			valueType: value === null ? "null" : typeof value,
			expected
		} });
	}
};

//#endregion
export { ConfigurationError, ValidationError };
