// @databricks/lakebase-eaa4c28d5584f767/dist/logger.js
//#region src/logger.ts
const LOGGER_METHODS = [
	"debug",
	"info",
	"warn",
	"error"
];
/**
* Check if the provided value is a Logger instance
*/
function isLogger(value) {
	if (typeof value !== "object" || value === null) return false;
	return LOGGER_METHODS.every((method) => method in value && typeof value[method] === "function");
}
/**
* Create a console-based logger from configuration
*/
function createConsoleLogger(config) {
	const noop = () => {};
	return {
		debug: config.debug ? console.debug.bind(console) : noop,
		info: config.info ? console.info.bind(console) : noop,
		warn: config.warn ? console.warn.bind(console) : noop,
		error: config.error ? console.error.bind(console) : noop
	};
}
/**
* Resolve logger configuration to a Logger instance
*
* - If Logger instance provided, return as-is
* - If LoggerConfig provided, create console-based logger
* - If undefined, create error-only logger (default)
*/
function resolveLogger(loggerConfig) {
	if (isLogger(loggerConfig)) return loggerConfig;
	if (loggerConfig && typeof loggerConfig === "object") return createConsoleLogger(loggerConfig);
	return createConsoleLogger({ error: true });
}

//#endregion
export { resolveLogger };
