// @databricks/lakebase-eaa4c28d5584f767/dist/pool-config.js
import { getUsernameSync, parsePoolConfig } from "./config.js";
import { initTelemetry } from "./telemetry.js";
import { createTokenRefreshCallback } from "./token-refresh.js";

//#region src/pool-config.ts
/**
* Map an SSL mode string to the corresponding `pg` SSL configuration.
*
* - `"require"` -- SSL enabled with certificate verification
* - `"prefer"`  -- SSL enabled without certificate verification (try SSL, accept any cert)
* - `"disable"` -- SSL disabled
*/
function mapSslConfig(sslMode) {
	switch (sslMode) {
		case "require": return { rejectUnauthorized: true };
		case "prefer": return { rejectUnauthorized: false };
		case "disable": return false;
	}
}
/**
* Get Lakebase connection configuration for PostgreSQL clients.
*
* Returns pg.PoolConfig with OAuth token authentication configured.
* Best used with pg.Pool directly or ORMs that accept pg.Pool instances (like Drizzle).
*
* For ORMs that need connection parameters (TypeORM, Sequelize), use getLakebaseOrmConfig() instead.
*
* Used internally by createLakebasePool().
*
* @param config - Optional configuration (reads from environment if not provided)
* @param telemetry - Optional pre-initialized telemetry (created internally if not provided)
* @param logger - Optional logger (silent if not provided)
* @returns PostgreSQL pool configuration with OAuth token refresh
*/
function getLakebasePgConfig(config, telemetry, logger) {
	const userConfig = config ?? {};
	const poolConfig = parsePoolConfig(userConfig);
	const username = getUsernameSync(userConfig);
	let passwordConfig;
	if (userConfig.password !== void 0) passwordConfig = userConfig.password;
	else if (poolConfig.endpoint) passwordConfig = createTokenRefreshCallback({
		userConfig,
		endpoint: poolConfig.endpoint,
		telemetry: telemetry ?? initTelemetry(),
		logger
	});
	return {
		host: poolConfig.host,
		port: poolConfig.port,
		user: username,
		database: poolConfig.database,
		password: passwordConfig,
		ssl: poolConfig.ssl ?? mapSslConfig(poolConfig.sslMode),
		max: poolConfig.max,
		idleTimeoutMillis: poolConfig.idleTimeoutMillis,
		connectionTimeoutMillis: poolConfig.connectionTimeoutMillis
	};
}
/**
* Get Lakebase connection configuration for ORMs that don't accept pg.Pool directly.
*
* Designed for ORMs like TypeORM and Sequelize that need connection parameters
* rather than a pre-configured pool instance.
*
* Returns connection config with field names compatible with common ORMs:
* - `username` instead of `user`
* - Simplified SSL config
* - Password callback support for OAuth token refresh
*
* @param config - Optional configuration (reads from environment if not provided)
* @returns ORM-compatible connection configuration
*
* @example
* ```typescript
* // TypeORM
* const dataSource = new DataSource({
*   type: 'postgres',
*   ...getLakebaseOrmConfig(),
*   entities: [User],
*   synchronize: true,
* });
*
* // Sequelize
* const sequelize = new Sequelize({
*   dialect: 'postgres',
*   ...getLakebaseOrmConfig(),
*   logging: false,
* });
* ```
*/
function getLakebaseOrmConfig(config) {
	const { user, password, ssl, ...pgConfig } = getLakebasePgConfig(config);
	return {
		...pgConfig,
		username: user,
		password,
		ssl: ssl ? typeof ssl === "boolean" ? ssl : { rejectUnauthorized: ssl.rejectUnauthorized } : false
	};
}

//#endregion
export { getLakebaseOrmConfig, getLakebasePgConfig };
