// @databricks/lakebase-eaa4c28d5584f767/dist/token-refresh.js
import { getWorkspaceClient } from "./config.js";
import { generateDatabaseCredential } from "./credentials.js";
import { SpanStatusCode } from "./telemetry.js";

//#region src/token-refresh.ts
const CACHE_BUFFER_MS = 120 * 1e3;
/** Fetch a fresh OAuth token from Databricks */
async function refreshToken(workspaceClient, endpoint) {
	const credential = await generateDatabaseCredential(workspaceClient, { endpoint });
	return {
		token: credential.token,
		expiresAt: new Date(credential.expire_time).getTime()
	};
}
/**
* Build the password callback with token caching, deduplication, and telemetry.
*
* The returned async function is called by `pg.Pool` each time a new connection
* is established. It caches OAuth tokens and deduplicates concurrent refresh
* requests so only one API call is made even under parallel connection creation.
*/
function createTokenRefreshCallback(deps) {
	let cachedToken;
	let tokenExpiresAt = 0;
	let workspaceClient = null;
	let refreshPromise = null;
	return async () => {
		if (!workspaceClient) try {
			workspaceClient = getWorkspaceClient(deps.userConfig);
		} catch (error) {
			deps.logger?.error("Failed to initialize workspace client: %O", error);
			throw error;
		}
		const now = Date.now();
		if (cachedToken && now < tokenExpiresAt - CACHE_BUFFER_MS) {
			const expiresIn = Math.round((tokenExpiresAt - now) / 1e3 / 60);
			deps.logger?.debug("Using cached OAuth token (expires in %d minutes at %s)", expiresIn, new Date(tokenExpiresAt).toISOString());
			return cachedToken;
		}
		const client = workspaceClient;
		if (!refreshPromise) refreshPromise = (async () => {
			const startTime = Date.now();
			try {
				const result = await deps.telemetry.tracer.startActiveSpan("lakebase.token.refresh", { attributes: { "lakebase.endpoint": deps.endpoint } }, async (span) => {
					const tokenResult = await refreshToken(client, deps.endpoint);
					span.setAttribute("lakebase.token.expires_at", new Date(tokenResult.expiresAt).toISOString());
					span.setStatus({ code: SpanStatusCode.OK });
					span.end();
					return tokenResult;
				});
				cachedToken = result.token;
				tokenExpiresAt = result.expiresAt;
				return cachedToken;
			} catch (error) {
				deps.logger?.error("Failed to fetch OAuth token: %O", {
					error,
					message: error instanceof Error ? error.message : String(error),
					endpoint: deps.endpoint
				});
				throw error;
			} finally {
				deps.telemetry.tokenRefreshDuration.record(Date.now() - startTime);
				refreshPromise = null;
			}
		})();
		return refreshPromise;
	};
}

//#endregion
export { createTokenRefreshCallback };
